Helpers/InsuranceHelper.php

```diff
@@ -0,0 +1,473 @@
+                    );
+                } catch (\Exception $e) {
+                    $this->logger->info('Impossible to create subscription Data : ', ['exception message :' => $e->getMessage(), 'Data' => $insuranceData]);
+                }
+            }
+
+
+        }
+        return $subscriptionArray;
+    }
+
+    /**
+     * @param Collection $itemsCollection
+     * @param array $subscriptionResult
+     * @param int $orderId // same for all subscription
+     * @param string $mode // same for all subscription
+     * @return \Alma\MonthlyPayments\Model\Insurance\Subscription[]
+     * @throws NoSuchEntityException
+     */
+    public function createDbSubscriptionArrayFromItemsAndApiResult(Collection $itemsCollection, array $subscriptionResult, string $mode): array
+    {
+        $dbSubscriptionArray = [];
+        /** @var \Magento\Sales\Model\Order\Invoice\Item $item */
+        foreach ($itemsCollection as $item) {
+            if (self::ALMA_INSURANCE_SKU !== $item->getSku()) {
+                continue;
+            }
+
+            /** @var \Alma\MonthlyPayments\Model\Insurance\Subscription $dbSubscription */
+            $orderItem = $item->getOrderItem();
+            $orderItemQty = (int)$item->getQty();
+            $orderItemInsuranceData = json_decode($orderItem->getData(self::ALMA_INSURANCE_DB_KEY), true);
+            $subscriptionResultContractData = [];
+            for ($i = 0; $i < $orderItemQty; $i++) {
+                foreach ($subscriptionResult as $key => $result) {
+                    if (array_search($orderItemInsuranceData['id'], $result)) {
+                        $subscriptionResultContractData = $result;
+                        unset($subscriptionResult[$key]);
+                        break;
+                    }
+                }
+                $dbSubscription = $this->subscriptionFactory->create();
+                $dbSubscription->setOrderId($orderItem->getOrderId());
+                $dbSubscription->setOrderItemId($orderItem->getItemId());
+                $dbSubscription->setName($orderItemInsuranceData['name']);
+                $dbSubscription->setSubscriptionId($subscriptionResultContractData['id']);
+                $dbSubscription->setSubscriptionAmount(intval($subscriptionResultContractData['amount']));
+                $dbSubscription->setContractId($orderItemInsuranceData['id']);
+                $dbSubscription->setCmsReference($subscriptionResultContractData['cms_reference']);
+                $dbSubscription->setLinkedProductName($orderItemInsuranceData['parent_name']);
+                $dbSubscription->setLinkedProductPrice($orderItemInsuranceData['parent_price']);
+                $dbSubscription->setSubscriptionState($subscriptionResultContractData['state']);
+                $dbSubscription->setSubscriptionMode($mode);
+                $dbSubscription->setCallbackUrl($this->getCallbackUrl());
+                $dbSubscriptionArray[] = clone $dbSubscription;
```

Comment:
Why do you need a clone here?

Reply:
object are send in reference in tab - Clone break reference